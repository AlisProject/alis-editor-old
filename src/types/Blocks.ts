export enum BlockType {
  Rule = 'Rule',
  Text = 'Text',
  Paragraph = 'Paragraph',
  Image = 'Image',
  Quote = 'Quote',
  Heading = 'Heading',
  Embed = 'Embed'
}

export interface Block {
  id: string
  type: BlockType
  payload?: any
  children?: Block[]
}

export interface RuleBlock extends Block {
  type: BlockType.Rule
}

export interface EmbedBlock extends Block {
  type: BlockType.Embed,
  payload: {
    src: string
  }
}

export interface ParagraphBlock extends Block {
  type: BlockType.Paragraph
  payload: {
    body: string
  }
}

export interface TextBlock extends Block {
  type: BlockType.Text
  payload: {
    body: string
  }
}

export interface ImageBlock extends Block {
  type: BlockType.Image
  payload: {
    src: string
    align?: string
  }
  children: Block[]
}

export interface HeadingBlock extends Block {
  type: BlockType.Heading
  payload: {
    size: 'h2' | 'h3'
    body: string
  }
  children: Block[]
}

export interface QuoteBlock extends Block {
  type: BlockType.Quote
  children: Block[]
}

/*
「詳細は [こちら](https://example.com) から確認できます」を再現する場合
[
  {
    type: 'Paragraph',
    children: [
      {
        type: 'Text',
        payload: {
          body: '詳細は'
        }
      },
      {
        type: 'Link',
        payload: {
          href: 'https://example.com'
        },
        children: [
          {
            type: 'Text',
            payload: {
              body: 'こちら'
            }
          }
        ]
      },
      {
        type: 'Text',
        payload: {
          body: 'から確認できます'
        }
      }
    ]
  }
]

*/
