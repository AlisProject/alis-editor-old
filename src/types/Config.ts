export interface AlisEditorConfig {
  articleId: string
  preview: boolean
  iframelyApikey: string
  uploadEndpoint: string
}
