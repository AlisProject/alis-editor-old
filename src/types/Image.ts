export interface ImageUploadRequest {
  article_image: string
}

export interface ImageUploadResponse {
  image_url: string
}
