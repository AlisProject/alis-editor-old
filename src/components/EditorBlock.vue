<template>
  <div
    class="block"
    :class="{
      'block--ondrag': onDrag
    }"
    @keydown="handleDisable"
    @click="handleClick"
    @drop.prevent="handleDrop"
    @dragover.prevent="handleDragover"
    @dragleave.prevent="handleDragLeave"
  >
    <component
      :is="`${typedBlock.type}Block`"
      :block="block"
      @update="handleUpdate"
      @delete="handleDelete"
      @append="handleAppendBlock"
    />
    <InsertButton
      v-if="typedBlock.type === 'Paragraph' && isActive"
      @disable="handleDisable"
      @append="handleAppendBlock"
      @upload="handleUpload"
    />
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import { Block, BlockType } from '../types/Blocks'
import RuleBlock from './RuleBlock.vue'
import ImageBlock from './ImageBlock.vue'
import ParagraphBlock from './ParagraphBlock.vue'
import HeadingBlock from './HeadingBlock.vue'
import QuoteBlock from './QuoteBlock.vue'
import InsertButton from './InsertButton.vue'
import { setTimeout } from 'timers'

export default Vue.extend({
  components: {
    ImageBlock,
    ParagraphBlock,
    HeadingBlock,
    QuoteBlock,
    RuleBlock,
    InsertButton
  },
  props: {
    block: Object
  },
  data(): { onDrag: boolean; isActive: boolean } {
    return {
      onDrag: false,
      isActive: true
    }
  },
  computed: {
    typedBlock(): Block {
      return this.block
    }
  },
  methods: {
    handleAppendBlock(type: BlockType) {
      this.$emit('append', type)
    },
    handleDisable() {
      setTimeout(() => {
        this.isActive = true
      }, 20)
    },
    handleClick() {
      this.isActive = true
    },
    handleDragover() {
      this.onDrag = true
    },
    handleDragLeave() {
      this.onDrag = false
    },
    handleUpload(event: Event) {
      this.$emit('upload', event)
    },
    handleDrop(event: Event) {
      event.preventDefault()
      event.stopPropagation()
      this.onDrag = false
      this.$emit('drop', event)
    },
    handleUpdate(event: any) {
      this.$emit('update', event)
    },
    handleDelete(event: any) {
      this.$emit('delete', event)
    }
  }
})
</script>

<style scoped>
.block {
  position: relative;
  border: solid 2px transparent;
}

.block.block--ondrag {
  border: dotted 2px #555;
}

.block.block--ondrag * {
  pointer-events: none;
}
</style>
