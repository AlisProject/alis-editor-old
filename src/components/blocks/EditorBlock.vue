<template>
  <div
    class="block"
    :class="{
      'block--ondrag': onDrag,
      'is-active': active
    }"
    @keydown="handleDisable"
    @click="handleClick"
    @drop.prevent="handleDrop"
    @dragover.prevent="handleDragover"
    @dragleave.prevent="handleDragLeave"
  >
    <template v-if="block.type">
      <component
        :is="`${block.type}Block`"
        :block="block"
        @input="handleInput"
        @update="handleUpdate"
        @delete="handleDelete"
        @append="handleAppendBlock"
        @addimageuri="handleAddImage"
      />
    </template>
    <InsertButton
      v-if="block.type === 'Paragraph' && active"
      @disable="handleDisable"
      @append="handleAppendBlock"
      @upload="handleUpload"
    />
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import { Block, BlockType } from '../../types/Blocks'
import RuleBlock from './RuleBlock.vue'
import ImageBlock from './ImageBlock.vue'
import ParagraphBlock from './ParagraphBlock.vue'
import HeadingBlock from './HeadingBlock.vue'
import QuoteBlock from './QuoteBlock.vue'
import InsertButton from '../InsertButton.vue'
import { setTimeout } from 'timers'

export default Vue.extend({
  components: {
    ImageBlock,
    ParagraphBlock,
    HeadingBlock,
    QuoteBlock,
    RuleBlock,
    InsertButton
  },
  props: {
    block: Object as () => Block,
    active: Boolean
  },
  data(): { onDrag: boolean, showInsertButton: boolean } {
    return {
      onDrag: false,
      showInsertButton: false
    }
  },
  methods: {
    handleDisable() {},
    handleClick() {
      this.$emit('active', this.block)
    },
    handleDragover() {
      this.onDrag = true
    },
    handleDragLeave() {
      this.onDrag = false
    },
    handleUpload(event: Event) {
      this.$emit('upload', event)
    },
    handleAddImage(src: string) {
      this.$emit('addimageuri', src)
    },
    handleDrop(event: Event) {
      event.preventDefault()
      event.stopPropagation()
      this.onDrag = false
      this.$emit('drop', event)
    },
    handleAppendBlock(type: BlockType) {
      this.$emit('append', type)
    },
    handleInput(event: any) {
      this.$emit('update', event)
    },
    handleUpdate(event: any) {
      this.$emit('update', event)
    },
    handleDelete(event: Block) {
      this.$emit('delete', event)
    }
  }
})
</script>

<style scoped>
.block {
  position: relative;
  border: solid 2px transparent;
}

.block.is-active {
  background: rgba(0, 0, 0, 0.01);
}

.block.block--ondrag {
  border: dotted 2px #555;
}

.block.block--ondrag * {
  pointer-events: none;
}

.mobile-space {
  display: none;
}

@media (max-width: 768px) {
  .mobile-space {
    display: block;
    padding-bottom: 60px;
  }

  .block.is-active {
    background: rgba(0, 0, 0, 0.05);
  }
}
</style>
