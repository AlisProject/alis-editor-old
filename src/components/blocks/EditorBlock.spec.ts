import { mount, config } from '@vue/test-utils'
import EditorBlock from './EditorBlock.vue'
import * as BlockTypes from '../../types/Blocks'
config.logModifiedComponents = false

describe('EditorBlock.vue', () => {
  test('renders correctly', () => {
    const block: BlockTypes.Block = {
      id: '123',
      type: BlockTypes.BlockType.Paragraph,
      payload: {
        body: '<p>foo</p>'
      }
    }
    const wrapper = mount(EditorBlock, {
      propsData: {
        block
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
