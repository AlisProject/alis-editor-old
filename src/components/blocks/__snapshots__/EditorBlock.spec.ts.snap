// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorBlock.vue renders correctly 1`] = `
<div class="block">
  <div class="wrapper paragraph">
    <div contenteditable="true" class="target paragraph">
      <p>foo</p>
    </div>
    <!---->
  </div>
  <!---->
</div>
`;
